### ISSUE TITLE
Require gradCAM for Siamese network

### ISSUE THREAD

**Issue by sutummala** [keras-team/keras#15647]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google colab
- TensorFlow installed from (source or binary): Yes
- TensorFlow version (use command below): 2.0 
- Python version: 3.x
- Bazel version (if compiling from source):
- GPU model and memory: P100
- Exact command to reproduce:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

In the below method, I want to access the last conv layer of base network using get_layer. However, I am getting graph disconnected error. 

def SiameseNetwork(input_shape):
    
    input_layer = tf.keras.Input(input_shape)

   # 1st 3D conv blocks, which involves, convolution, BN, activation and pooling 
    x_1 = tf.keras.layers.Conv3D(32, (3,3,3), strides = (1,1,1), padding = 'same', kernel_regularizer = 'L2', name = 'conv3d_1')(input_layer)
    x_1_bn = tf.keras.layers.BatchNormalization(axis = -1)(x_1)
    x_1_bn_ac = tf.keras.layers.Activation('relu')(x_1_bn)

    x_2 = tf.keras.layers.Conv3D(32, (3,3,3), strides = (1,1,1), padding = 'same', kernel_regularizer = 'L2', name = 'conv3d_2')(x_1_bn_ac)
    x_2_bn = tf.keras.layers.BatchNormalization(axis = -1)(x_2)
    x_2_bn_ac = tf.keras.layers.Activation('relu')(x_2_bn)
    x_2_bn_ac_pooling = tf.keras.layers.MaxPooling3D(strides = (2, 2, 2))(x_2_bn_ac)
              
    # 2nd 3D conv block, which involves, convolution, BN, activation and pooling 
    x_3 = tf.keras.layers.Conv3D(64, (3,3,3), strides = (1,1,1), padding = 'same', kernel_regularizer = 'L2', name = 'conv3d_3')(x_2_bn_ac_pooling)
    x_3_bn = tf.keras.layers.BatchNormalization(axis = -1)(x_3)
    x_3_bn_ac = tf.keras.layers.Activation('relu')(x_3_bn)

    x_4 = tf.keras.layers.Conv3D(64, (3,3,3), strides = (1,1,1), padding = 'same', kernel_regularizer = 'L2', name = 'conv3d_4')(x_3_bn_ac)
    x_4_bn = tf.keras.layers.BatchNormalization(axis = -1)(x_4)
    x_4_bn_ac = tf.keras.layers.Activation('relu')(x_4_bn)
    x_4_bn_ac_pooling = tf.keras.layers.MaxPooling3D(strides = (2, 2, 2))(x_4_bn_ac)
    
    # 3rd 3D conv block, which involves, convolution, BN, activation and pooling 
    x_5 = tf.keras.layers.Conv3D(256, (3,3,3), strides = (1,1,1), padding = 'same', kernel_regularizer = 'L2', name = 'conv3d_5')(x_4_bn_ac_pooling)
    x_5_bn = tf.keras.layers.BatchNormalization(axis = -1)(x_5)
    x_5_bn_ac = tf.keras.layers.Activation('relu')(x_5_bn)
    x_5_pooling = tf.keras.layers.MaxPooling3D(strides = (2, 2, 2))(x_5_bn_ac)       
    
    gap_layer = tf.keras.layers.GlobalAveragePooling3D()(x_5_pooling) # Global average pooling layer
    #model.add(tf.keras.layers.Dropout(0.3))
    embeddings = tf.keras.layers.Dense(1024, activation = 'relu', kernel_regularizer = 'L2')(gap_layer)
    
    encoding_model = tf.keras.Model(inputs = input_layer, outputs =  embeddings, name = 'base_3dcnn')

    # Siamese model starts

    moving_input = tf.keras.Input(input_shape)
    ref_input = tf.keras.Input(input_shape)

    encoded_moving  = encoding_model(moving_input)
    encoded_ref = encoding_model(ref_input)

    L1_layer = tf.keras.layers.Lambda(lambda tensors:K.abs(tensors[0] - tensors[1]))
    #L2_layer = tf.keras.layers.Lambda(lambda tensors:K.l2_normalize((tensors[0] - tensors[1]), axis = 1))

    L1_distance = L1_layer([encoded_moving, encoded_ref]) # L1-norm
    #dot_product = tf.keras.layers.dot([encoded_moving, encoded_ref], axes = 1, normalize = False)
    #L2_distance = L2_layer([encoded_moving, encoded_ref]) # L2-norm or Euclidean Norm
    #L2_distance = tf.keras.layers.Lambda(euclidean_distance, output_shape=eucl_dist_output_shape)([encoded_moving, encoded_ref])

    prediction = tf.keras.layers.Dense(1, activation='sigmoid')(L1_distance)
    siamesenet = tf.keras.Model(inputs = [moving_input, ref_input], outputs = prediction, name = 'siamese_3dmodel')
    
    return siamesenet, encoding_model

**Describe the current behavior**.

Below code gives the error. 

siamese_model.layers[-1].activation = None

last_conv_layer_name = 'conv3d_5'
last_layer = siamese_model.get_layer('base_3dcnn').get_layer(last_conv_layer_name)
s_model = tf.keras.models.Model(
        [siamese_model.inputs], [last_layer.output, siamese_model.output]
    )
**Describe the expected behavior**.

I want to compute the gradients with respect to the last conv layer of the base model. I have two inputs since it is a twin network, one of it is the reference. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by sanatmpa1** [comment]
@sutummala,

Can you share a complete standalone code or colab gist to reproduce the issue? Thanks!

**Comment by sutummala** [comment]
[https://colab.research.google.com/drive/1RYG6UzfEDt3vSIVikfkTMln-0vloRFP0?authuser=3#scrollTo=CjiNmvUxxzIQ](url)

**Comment by sutummala** [comment]
[https://gist.github.com/sutummala/4f298d29bc47fc7894b00704d78ca810](url)

**Comment by sanatmpa1** [comment]
@sutummala,

Can you take a look at this links for similar [issue1](https://github.com/keras-team/keras/issues/11151), [issue2](https://stackoverflow.com/questions/56589726/valueerror-graph-disconnected-cannot-obtain-value-for-tensor-tensor-the-foll), [issue3](https://stackoverflow.com/questions/51522848/graph-disconnected-cannot-obtain-value-for-tensor-tensor-input-keras-python/51523029) about graph disconnected errors, and let me know if it can be of any help? Thanks
